File: Tests/ProxyClassifierTests.swift
import XCTest
import os

@testable import MihomoMeter

final class ProxyClassifierTests: XCTestCase {
  private let classifier = ProxyClassifier(
    catalog: ProxyCatalog(
      typesByName: [
        "DIRECT": "Direct",
        "REJECT": "Reject",
        "REJECT-DROP": "RejectDrop",
        "Synthetic Group": "Selector",
        "Synthetic Proxy": "Vmess",
      ]
    )
  )

  func testClassifiesConcreteLeafWithoutUsingOuterGroup() {
    XCTAssertEqual(
      classifier.classify(chains: ["Synthetic Proxy", "Synthetic Group"]),
      ProxyClassification(category: .proxy, unknownReason: nil)
    )
  }

  func testClassifiesEverySupportedConcreteProxyType() {
    let supportedTypes = [
      "AnyTLS",
      "Http",
      "Hysteria",
      "Hysteria2",
      "Shadowsocks",
      "ShadowsocksR",
      "Snell",
      "Socks5",
      "Ssh",
      "Trojan",
      "Tuic",
      "Vless",
      "Vmess",
      "WireGuard",
    ]

    for type in supportedTypes {
      let classifier = ProxyClassifier(
        catalog: ProxyCatalog(typesByName: ["Synthetic Proxy": type])
      )
      XCTAssertEqual(
        classifier.classify(chains: ["Synthetic Proxy"]),
        ProxyClassification(category: .proxy, unknownReason: nil),
        "\(type) 应归类为 Proxy"
      )
    }
  }

  func testKeepsDirectAndRejectSeparateFromProxy() {
    XCTAssertEqual(
      classifier.classify(chains: ["DIRECT"]).category,
      .direct
    )
    XCTAssertEqual(
      classifier.classify(chains: ["REJECT"]).category,
      .reject
    )
    XCTAssertEqual(
      classifier.classify(chains: ["REJECT-DROP"]).category,
      .reject
    )
  }

  func testDoesNotGuessMissingOrAmbiguousLeaf() {
    XCTAssertEqual(
      classifier.classify(chains: []).unknownReason,
      .emptyChain
    )
    XCTAssertEqual(
      classifier.classify(chains: ["Not In Catalog"]).unknownReason,
      .missingCatalogEntry
    )
    XCTAssertEqual(
      classifier.classify(chains: ["Synthetic Group"]).unknownReason,
      .ambiguousProxyType
    )
  }

  func testInvokesInjectedResolverOnlyAfterCatalogHitWithNativeClassification() {
    let invocations = OSAllocatedUnfairLock(
      initialState: [(rawType: String, classification: ProxyClassification)]()
    )
    let classifier = ProxyClassifier(
      catalog: ProxyCatalog(typesByName: ["Synthetic Proxy": "Vmess"]),
      resolveProxyType: { rawType, nativeClassification in
        invocations.withLock {
          $0.append((rawType: rawType, classification: nativeClassification))
        }
        return nativeClassification
      }
    )

    _ = classifier.classify(chains: [])
    _ = classifier.classify(chains: ["Missing"])
    XCTAssertEqual(
      classifier.classify(chains: ["Synthetic Proxy"]),
      ProxyClassification(category: .proxy, unknownReason: nil)
    )
    XCTAssertEqual(invocations.withLock { $0.count }, 1)
    XCTAssertEqual(invocations.withLock { $0.first?.rawType }, "Vmess")
    XCTAssertEqual(
      invocations.withLock { $0.first?.classification },
      ProxyClassification(category: .proxy, unknownReason: nil)
    )
  }

  func testLazyResolverControlsWhetherNativeClassificationIsEvaluated() {
    let invocations = OSAllocatedUnfairLock(initialState: [String]())
    let sharedClassifier = ProxyClassifier(
      catalog: ProxyCatalog(typesByName: ["Synthetic": "Selector"]),
      resolveProxyTypeLazily: { rawType, _ in
        invocations.withLock { $0.append(rawType) }
        return ProxyClassification(category: .proxy, unknownReason: nil)
      }
    )
    XCTAssertEqual(
      sharedClassifier.classify(chains: ["Synthetic"]),
      ProxyClassification(category: .proxy, unknownReason: nil)
    )

    let fallbackClassifier = ProxyClassifier(
      catalog: ProxyCatalog(typesByName: ["Synthetic": "Selector"]),
      resolveProxyTypeLazily: { _, nativeFallback in nativeFallback() }
    )
    XCTAssertEqual(
      fallbackClassifier.classify(chains: ["Synthetic"]),
      ProxyClassification(category: .unknown, unknownReason: .ambiguousProxyType)
    )
    XCTAssertEqual(invocations.withLock { $0 }, ["Selector"])
  }
}

File: platform/windows/MihomoMeter.Windows.Tests/ProxyClassifierTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MihomoMeter.Windows.Core.Domain;

namespace MihomoMeter.Windows.Tests;

[TestClass]
public sealed class ProxyClassifierTests
{
    private static readonly ProxyClassifier Classifier = new(new ProxyCatalog(
        new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["DIRECT"] = "Direct",
            ["REJECT"] = "Reject",
            ["REJECT-DROP"] = "RejectDrop",
            ["Synthetic Group"] = "Selector",
            ["Synthetic Proxy"] = "Vmess",
        }));

    [TestMethod]
    public void ClassifiesConcreteLeafWithoutUsingOuterGroup()
    {
        var result = Classifier.Classify(["Synthetic Proxy", "Synthetic Group"]);

        Assert.AreEqual(TrafficCategory.Proxy, result.Category);
        Assert.IsNull(result.UnknownReason);
    }

    [TestMethod]
    public void ClassifiesEverySupportedConcreteProxyType()
    {
        string[] supportedTypes =
        [
            "AnyTLS",
            "Http",
            "Hysteria",
            "Hysteria2",
            "Shadowsocks",
            "ShadowsocksR",
            "Snell",
            "Socks5",
            "Ssh",
            "Trojan",
            "Tuic",
            "Vless",
            "Vmess",
            "WireGuard",
        ];

        foreach (var type in supportedTypes)
        {
            var classifier = new ProxyClassifier(new ProxyCatalog(
                new Dictionary<string, string>(StringComparer.Ordinal)
                {
                    ["Synthetic Proxy"] = type,
                }));

            Assert.AreEqual(
                new ProxyClassification(TrafficCategory.Proxy),
                classifier.Classify(["Synthetic Proxy"]),
                $"{type} 应归类为 Proxy");
        }
    }

    [TestMethod]
    public void KeepsDirectAndRejectSeparateFromProxy()
    {
        Assert.AreEqual(TrafficCategory.Direct, Classifier.Classify(["DIRECT"]).Category);
        Assert.AreEqual(TrafficCategory.Reject, Classifier.Classify(["REJECT"]).Category);
        Assert.AreEqual(TrafficCategory.Reject, Classifier.Classify(["REJECT-DROP"]).Category);
    }

    [TestMethod]
    public void DoesNotGuessMissingOrAmbiguousLeaf()
    {
        Assert.AreEqual(
            UnknownTrafficReason.EmptyChain,
            Classifier.Classify([]).UnknownReason);
        Assert.AreEqual(
            UnknownTrafficReason.MissingCatalogEntry,
            Classifier.Classify(["Missing"]).UnknownReason);
        Assert.AreEqual(
            UnknownTrafficReason.AmbiguousProxyType,
            Classifier.Classify(["Synthetic Group"]).UnknownReason);
    }

    [TestMethod]
    public void InvokesInjectedResolverOnlyAfterCatalogHitWithNativeClassification()
    {
        var invocations = new List<(string RawType, ProxyClassification Classification)>();
        var classifier = new ProxyClassifier(
            new ProxyCatalog(new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["Synthetic Proxy"] = "Vmess",
            }),
            (rawType, nativeClassification) =>
            {
                invocations.Add((rawType, nativeClassification));
                return nativeClassification;
            });

        _ = classifier.Classify([]);
        _ = classifier.Classify(["Missing"]);
        Assert.AreEqual(
            new ProxyClassification(TrafficCategory.Proxy),
            classifier.Classify(["Synthetic Proxy"]));
        Assert.AreEqual(1, invocations.Count);
        Assert.AreEqual("Vmess", invocations[0].RawType);
        Assert.AreEqual(
            new ProxyClassification(TrafficCategory.Proxy),
            invocations[0].Classification);
    }

    [TestMethod]
    public void LazyResolverControlsWhetherNativeClassificationIsEvaluated()
    {
        var invocations = new List<string>();
        var sharedClassifier = new ProxyClassifier(
            new ProxyCatalog(new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["Synthetic"] = "Selector",
            }),
            (LazyProxyTypeResolver)((rawType, _) =>
            {
                invocations.Add(rawType);
                return new ProxyClassification(TrafficCategory.Proxy);
            }));
        Assert.AreEqual(
            new ProxyClassification(TrafficCategory.Proxy),
            sharedClassifier.Classify(["Synthetic"]));

        var fallbackClassifier = new ProxyClassifier(
            new ProxyCatalog(new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["Synthetic"] = "Selector",
            }),
            (LazyProxyTypeResolver)((_, nativeFallback) => nativeFallback()));
        Assert.AreEqual(
            new ProxyClassification(
                TrafficCategory.Unknown,
                UnknownTrafficReason.AmbiguousProxyType),
            fallbackClassifier.Classify(["Synthetic"]));
        CollectionAssert.AreEqual(new[] { "Selector" }, invocations);
    }
}
